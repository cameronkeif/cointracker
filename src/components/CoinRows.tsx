import React, {
  useEffect, useState, Fragment, ChangeEvent, KeyboardEvent,
} from 'react';
import TextField from '@material-ui/core/TextField';
import IconButton from '@material-ui/core/IconButton';
import { green, grey } from '@material-ui/core/colors';
import AddCircleIcon from '@material-ui/icons/AddCircle';
import Autocomplete, { AutocompleteRenderInputParams } from '@material-ui/lab/Autocomplete';
import request from 'superagent';

import CoinRow from './CoinRow';
import api from '../api';
import { CoinItem, CoinOption } from '../react-app-env';
import spinnerIcon from '../assets/spinner.png';

import coinOptions from '../data/coin-options.json';

const CoinRows = () => {
  const [symbols, setSymbols] = useState<Set<string>>(new Set());
  const [selectedCoin, setSelectedCoin] = useState<null | CoinOption>(null);
  const [coins, setCoins] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [activeTimerId, setActiveTimerId] = useState<null | NodeJS.Timeout>(null);

  const addCoin = () => {
    if (selectedCoin === null) {
      return;
    }

    const updatedSymbols = new Set(symbols);
    updatedSymbols.add(selectedCoin.symbol);

    setSymbols(updatedSymbols);
    setSelectedCoin(null);

    if (activeTimerId) {
      clearInterval(activeTimerId);
    }
  };

  const removeCoin = (symbol: string) => {
    const updatedSymbols = new Set(symbols);
    updatedSymbols.delete(symbol);
    setSymbols(updatedSymbols);

    if (activeTimerId) {
      clearInterval(activeTimerId);
    }

    if (updatedSymbols.size === 0) {
      setCoins([]);
    }
  };

  useEffect(() => {
    if (symbols.size === 0) {
      return;
    }

    api.getCoinData(symbols)
      .then((res: request.Response) => {
        setCoins(res.body.data.coins);
        setIsLoading(false); // Clear initial loading state.
      });
    setActiveTimerId(setInterval(() => {
      setIsLoading(true);
      api.getCoinData(symbols)
        .then((res: request.Response) => {
          setCoins(res.body.data.coins);
          setIsLoading(false);
        });
    }, 15000));
  }, [symbols]);
  const coinRows = coins.map((coin: CoinItem) => (
    <CoinRow
      coin={coin}
      onRemove={removeCoin}
      key={`coin-row-${coin.uuid}`}
    />
  ));

  const addButtonColor = selectedCoin ? green[500] : grey[500];
  return (
    <>
      <table className="coin-table">
        <thead>
          <tr>
            <th>Name</th>
            <th>Symbol</th>
            <th>Price ($)</th>
            <th>Change (24hr)</th>
          </tr>
        </thead>
        <tbody>
          {coinRows}
        </tbody>
      </table>
      <div style={{
        display: 'inline-flex', alignItems: 'center', width: 'auto', marginTop: 20,
      }}
      >
        <Autocomplete
          id="combo-box-demo"
          options={coinOptions.filter((coinOption: CoinOption) => !symbols.has(coinOption.symbol))}
          style={{ width: 300, margin: 'auto', display: 'inline-block' }}
          renderInput={(params: AutocompleteRenderInputParams) => (
            <TextField
              {...params}
              label="Add coin"
              size="small"
              variant="outlined"
            />
          )}
          getOptionLabel={(coinOption: CoinOption) => `${coinOption.name} (${coinOption.symbol})`}
          onChange={(
            event: ChangeEvent<{}>,
            newSelectedCoin: { name: string; symbol: string; } | null,
          ) => { setSelectedCoin(newSelectedCoin); }}
          onKeyDown={(event: KeyboardEvent<HTMLDivElement>) => {
            if (event.key === 'Enter') {
              addCoin();
            }
          }}
          value={selectedCoin}
        />
        <IconButton
          disabled={!selectedCoin}
          onClick={addCoin}
        >
          <AddCircleIcon
            style={{ color: addButtonColor, cursor: 'pointer' }}
            fontSize="large"
          />
        </IconButton>
      </div>
      {isLoading && symbols.size > 0 && <div><img src={spinnerIcon} className="loading" alt="Loading" /></div>}
    </>
  );
};

export default CoinRows;
