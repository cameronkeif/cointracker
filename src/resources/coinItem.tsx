import { CoinItem } from '../react-app-env';

const coinItem: CoinItem = {
  change: -2.06,
  symbol: 'BTC',
  name: 'Bitcoin',
  uuid: 'Qwsogvtv82FCd',
  price: '55099.8441599246',
  iconUrl: 'https://cdn.coinranking.com/bOabBYkcX/bitcoin_btc.svg',
};

export default coinItem;
