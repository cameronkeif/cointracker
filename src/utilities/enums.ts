export enum TableSortType {
  Name = 'name',
  Symbol = 'symbol',
  Price = 'price',
  Change = 'change',
}

export enum TableSortDirection {
  Ascending,
  Descending,
}
