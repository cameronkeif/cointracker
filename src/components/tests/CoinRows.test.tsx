import React from 'react';
import nock from 'nock';
import { shallow, mount } from 'enzyme';

import CoinRows, { TableSort } from '../CoinRows';
import coinItem from '../../resources/coinItem';
import { TableSortType, TableSortDirection } from '../../utilities/enums';

const flushPromises = () => new Promise(setImmediate);

const generateMockRequest = () => nock('https://api.coinranking.com')
  .get('/v1/public/coins')
  .query(true)
  .reply(200, { data: { coins: [coinItem] } });

describe('CoinRows.tsx', () => {
  const realUseState = React.useState;

  beforeEach(() => {
    React.useState = realUseState;
  });

  it('renders correctly - empty', () => {
    expect(shallow(<CoinRows />)).toMatchSnapshot();
  });

  it('renders correctly - loading', () => {
    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(), {}])
      .mockReturnValueOnce([[], {}])
      .mockReturnValueOnce([true, {}])
      .mockReturnValueOnce([null, {}])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, {}]);

    expect(shallow(<CoinRows />)).toMatchSnapshot();
  });

  it('renders correctly - populated', async () => {
    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC']), {}])
      .mockReturnValueOnce([[coinItem], {}])
      .mockReturnValueOnce([false, {}])
      .mockReturnValueOnce([null, {}])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, {}]);

    expect(shallow(<CoinRows />)).toMatchSnapshot();
  });

  it('calls API on mount', async () => {
    const pending = generateMockRequest();

    // Initialize state for simplicity.
    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC']), jest.fn()])
      .mockReturnValueOnce([[coinItem], jest.fn()])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([null, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, jest.fn()])
      .mockReturnValueOnce([null, jest.fn()]);

    mount(<CoinRows />);
    await flushPromises();
    expect(pending.isDone()).toBe(true);
  });

  const testAdd = (
    shouldClearIntervalBeCalled: boolean,
    setSymbols: Function,
  ) => {
    generateMockRequest();

    jest.spyOn(window, 'clearInterval');

    const addButton = mount(<CoinRows />).find('button[data-id="add-btn"]');
    addButton.simulate('click');
    expect(setSymbols).toHaveBeenCalledWith(new Set(['BTC', 'DOGE']));
    if (shouldClearIntervalBeCalled) {
      expect(clearInterval).toHaveBeenCalled();
    } else {
      expect(clearInterval).not.toHaveBeenCalled();
    }
  };

  it('Handles adding a coin', () => {
    const setSymbols = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC']), setSymbols])
      .mockReturnValueOnce([[coinItem], jest.fn()])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([5, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, jest.fn()])
      .mockReturnValueOnce([{ name: 'Dogecoin', symbol: 'DOGE' }, jest.fn()]);

    testAdd(true, setSymbols);
  });

  it('Handles adding a coin (no interval set)', () => {
    const setSymbols = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC']), setSymbols])
      .mockReturnValueOnce([[coinItem], jest.fn()])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([null, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, jest.fn()])
      .mockReturnValueOnce([{ name: 'Dogecoin', symbol: 'DOGE' }, jest.fn()]);

    testAdd(false, setSymbols);
  });

  const testDelete = (
    shouldClearIntervalBeCalled: boolean,
    setSymbols: Function,
    expectedSymbols: Set,
    setCoins: Function,
    shouldTestCoins: boolean,
  ) => {
    generateMockRequest();

    jest.spyOn(window, 'clearInterval');

    const deleteButton = mount(<CoinRows />).find('.coin-row .MuiSvgIcon-root');
    deleteButton.first().simulate('click');
    expect(setSymbols).toHaveBeenCalledWith(expectedSymbols);

    if (shouldClearIntervalBeCalled) {
      expect(clearInterval).toHaveBeenCalled();
    } else {
      expect(clearInterval).not.toHaveBeenCalled();
    }

    if (shouldTestCoins) {
      expect(setCoins).toHaveBeenCalledWith([]);
    } else {
      expect(setCoins).not.toHaveBeenCalled();
    }
  };

  it('Handles deleting a coin', () => {
    const coinItem2 = {
      ...coinItem,
      name: 'Dogecoin',
      symbol: 'DOGE',
      uuid: 'Qwsogvtv82FCe',
    };

    const setSymbols = jest.fn();
    const setCoins = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC', 'DOGE']), setSymbols])
      .mockReturnValueOnce([[coinItem, coinItem2], setCoins])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([5, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, jest.fn()])
      .mockReturnValueOnce([null, jest.fn()]);

    testDelete(true, setSymbols, new Set(['DOGE']), setCoins, false);
  });

  it('Handles deleting a coin (no coins after delete) (no interval set)', () => {
    const setSymbols = jest.fn();
    const setCoins = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC']), setSymbols])
      .mockReturnValueOnce([[coinItem], setCoins])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([null, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, jest.fn()])
      .mockReturnValueOnce([null, jest.fn()]);

    testDelete(false, setSymbols, new Set(), setCoins, true);
  });

  const testSort: void = (headerIndex: number, expectedTableSort: TableSort, setSortType: Function) => {
    const component = mount(<CoinRows />);

    const nameHeader = component.find('th > div');

    nameHeader.at(headerIndex).simulate('keydown', { key: 'Enter' });
    expect(setSortType.mock.calls[0][0]).toEqual(expectedTableSort);

    nameHeader.at(headerIndex).simulate('click');
    expect(setSortType.mock.calls[0][0]).toEqual(expectedTableSort);

    // Clear and retry with keypress
    jest.clearAllMocks();

    nameHeader.at(headerIndex).simulate('keydown', { key: 'Enter' });
    expect(setSortType.mock.calls[0][0]).toEqual(expectedTableSort);
  };

  it('Handles changing the sort order (same sort field)', () => {
    const coinItem2 = {
      ...coinItem,
      name: 'Dogecoin',
      symbol: 'DOGE',
      uuid: 'Qwsogvtv82FCe',
    };

    const setSymbols = jest.fn();
    const setCoins = jest.fn();

    const setSortType = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC', 'DOGE']), setSymbols])
      .mockReturnValueOnce([[coinItem, coinItem2], setCoins])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([5, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, setSortType])
      .mockReturnValueOnce([null, jest.fn()]);

    testSort(0, { sortDirection: TableSortDirection.Descending, sortType: TableSortType.Name }, setSortType);
  });

  it('Handles changing the sort order (different sort field - default ascending)', () => {
    const coinItem2 = {
      ...coinItem,
      name: 'Dogecoin',
      symbol: 'DOGE',
      uuid: 'Qwsogvtv82FCe',
    };

    const setSymbols = jest.fn();
    const setCoins = jest.fn();

    const setSortType = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC', 'DOGE']), setSymbols])
      .mockReturnValueOnce([[coinItem, coinItem2], setCoins])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([5, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, setSortType])
      .mockReturnValueOnce([null, jest.fn()]);

    testSort(1, { sortDirection: TableSortDirection.Ascending, sortType: TableSortType.Symbol }, setSortType);
  });

  it('Handles changing the sort order (different sort field - default descending)', () => {
    const coinItem2 = {
      ...coinItem,
      name: 'Dogecoin',
      symbol: 'DOGE',
      uuid: 'Qwsogvtv82FCe',
    };

    const setSymbols = jest.fn();
    const setCoins = jest.fn();

    const setSortType = jest.fn();

    React.useState = jest.fn()
      .mockReturnValueOnce([new Set(['BTC', 'DOGE']), setSymbols])
      .mockReturnValueOnce([[coinItem, coinItem2], setCoins])
      .mockReturnValueOnce([false, jest.fn()])
      .mockReturnValueOnce([5, jest.fn()])
      .mockReturnValueOnce([{ sortType: TableSortType.Name, sortDirection: TableSortDirection.Ascending }, setSortType])
      .mockReturnValueOnce([null, jest.fn()]);

    testSort(2, { sortDirection: TableSortDirection.Descending, sortType: TableSortType.Price }, setSortType);
  });
});
