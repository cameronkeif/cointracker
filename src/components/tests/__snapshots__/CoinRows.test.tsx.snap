// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinRows renders correctly - empty 1`] = `
<DocumentFragment>
  <table
    class="coin-table"
  >
    <thead>
      <tr>
        <th>
          <div
            role="button"
            style="overflow: auto; white-space: nowrap;"
            tabindex="0"
          >
            <span>
              Name
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 14l5-5 5 5z"
                />
              </svg>
            </span>
          </div>
        </th>
        <th>
          <div
            role="button"
            style="overflow: auto; white-space: nowrap;"
            tabindex="0"
          >
            <span>
              Symbol
            </span>
          </div>
        </th>
        <th>
          <div
            role="button"
            style="overflow: auto; white-space: nowrap;"
            tabindex="0"
          >
            <span>
              Price
            </span>
          </div>
        </th>
        <th>
          <div
            role="button"
            style="overflow: auto; white-space: nowrap;"
            tabindex="0"
          >
            <span>
              Change (24 hours)
            </span>
          </div>
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div
    style="display: inline-flex; align-items: center; width: auto; margin-top: 20px;"
  >
    <div
      aria-expanded="false"
      class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
      role="combobox"
      style="width: 300px; margin: auto; display: inline-block;"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        data-id="coin-input"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
          data-shrink="false"
          for="combo-box-demo"
          id="combo-box-demo-label"
        >
          Add coin
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        >
          <input
            aria-autocomplete="list"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            id="combo-box-demo"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Clear"
              class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
              tabindex="-1"
              title="Clear"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Add coin
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
      data-id="add-btn"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          style="color: rgb(158, 158, 158); cursor: pointer;"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
